keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,474 @@
+   graceful termination logic.
+
+This implies the following:
+- Pods of a workload could implement checkpointing as part of their graceful
+  termination.
+- The resources from these Pods are not immediately available and they could
+  be arbitrarily delayed.
+- While Pods terminate, a ClusterQueue's quota could be oversubscribed.
+
+Oversubscription could be mitigated by watching the running pods but this will
+be left for future work.
+
+#### Why no control to opt-out a ClusterQueue from preemption
+
+In a cohort, some ClusterQueue could have high priority Workloads running, so
+it might be desired not to disturb them.
+
+However, this could be achieved by two means:
+- Restricting other ClusterQueues to only have workloads with lower priority,
+  via [QuotaScopes](https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-scopes)
+- Configuring the ClusterQueue with high priority Workloads to never borrow
+  (through `.quota.max`), while owning a big part or all of the quota for the
+  cohort.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+### Increased admission latency
+
+TODO
+
+## Design Details
+
+The proposal consists of new API fields and a preemption algorithm.
+
+### API
+
+#### ClusterQueue changes
+
+The new API fields in ClusterQueue describe how to influence the selection
+of Workloads to preempt.
+
+```golang
+type ClusterQueueSpec struct {
+  ...
+  // preemption describes policies to preempt Workloads from ClusterQueues
+  // borrowing resources when this ClusterQueue has pending Workloads and the
+  // min quota is unmet.
+  preemption ClusterQueuePreemption
+}
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish before issuing preemptions,
+  // since the first time that the ClusterQueue has pending workloads that would
+  // fit under the min quota. This time is present in the timestamp for the
+  // PendingWorkloadsUnderMinQuota condition.
+  // If null, it means to wait indefinitely.
+  WaitBeforePreemptionSeconds *int64
+}
+```
+
+### Workload changes
+
+The new API field in Workload describe what to do in the case of a preemption.
+
+```golang
+type WorkloadSpec struct {
+  ...
+  // onPreemption describes what to do with the workload when it's preempted.
+  // Options are:
+  //
+  // - Requeue (default): The Workload is put back in the queue for admission.
+  //   Note that the original .metadata.CreationTimestamp would be used for
+  //   sorting this workload.
+  // - Terminate: The workload gets a Finished condition with reason Preempted.
```

Comment:
Will this be propagated back to Job status ?

Reply:
I wanted to, but it's rather hard. See mitigations.